### ISSUE TITLE
Unhandled 'etimedout' in 'gen_tcp_socket'

### ISSUE THREAD

**Issue by x0id** [erlang/otp#7608]
**Describe the bug**
We caught the following error in OTP 24:
```
[{supervisor,{local,my_supervisor}},
 {errorContext,child_terminated},
 {reason,
     {{{case_clause,etimedout},
       [{gen_tcp_socket,handle_recv_error,4,
            [{file,"gen_tcp_socket.erl"},{line,2235}]},
        {gen_statem,loop_state_callback,11,
            [{file,"gen_statem.erl"},{line,1205}]},
        {proc_lib,init_p_do_apply,3,[{file,"proc_lib.erl"},{line,226}]}]},
      {gen_statem,call,[<0.28057.989>,close,infinity]}}},
```

**To Reproduce**
In Server:
```
iptables -I OUTPUT -p tcp --sport 8888 --tcp-flags ALL ACK -j DROP
nc -l 8888
```
In Client:
```
erl -kernel inet_backend socket
```
run code:
```
{ok, Socket} = gen_tcp:connect(Host, 8888, [{debug, true}]),
gen_tcp:send(Socket, <<"test">>),
receive
    Msg ->
        io:format("~nmsg: ~p~n~n", [Msg])
end,
gen_tcp:close(Socket).
```

**Expected behavior**
`case_clause` error should not happen in `gen_tcp_socket`

**Affected versions**
24.3.4.6-slim

**Additional context**

Per logs the error happened when app called `gen_tcp:close(Socket)` soon after receiving message `{tcp_error, Socket, etimedout}`.

The error message with the stack was:
```
gen_tcp_socket call failed: 
      Call:  close
      Class: exit
      Error: {{{case_clause,etimedout},
               [{gen_tcp_socket,handle_recv_error,4,
                                [{file,"gen_tcp_socket.erl"},{line,2235}]},
                {gen_statem,loop_state_callback,11,
                            [{file,"gen_statem.erl"},{line,1205}]},
                {proc_lib,init_p_do_apply,3,
                          [{file,"proc_lib.erl"},{line,226}]}]},
              {gen_statem,call,[<0.15909.999>,close,infinity]}}
      Stack: [{gen,do_call,4,[{file,"gen.erl"},{line,220}]},
              {gen_statem,call_dirty,4,[{file,"gen_statem.erl"},{line,703}]},
              {gen_tcp_socket,call,2,
                              [{file,"gen_tcp_socket.erl"},{line,1134}]},
              {gen_tcp_socket,close_server,1,
                              [{file,"gen_tcp_socket.erl"},{line,474}]},
              {gen_tcp_socket,close,1,
                              [{file,"gen_tcp_socket.erl"},{line,469}]},
```

`gen_statem` error message:
```
{
  "callback_mode": "handle_event_function",
  "client_info": "undefined",
  "file": "gen_statem.erl",
  "label": "{gen_statem,terminate}",
  "level": "error",
  "line": 2424,
  "log": "",
  "mfa": "gen_statem:error_info/7",
  "modules": "[gen_tcp_socket]",
  "name": "<0.17008.1001>",
  "pid": "<0.17008.1001>",
  "postponed": "",
  "queue": "[{info,{'$socket',{'$socket',#Ref<0.3186442261.3805675525.237627>},select,#Ref<0.3186442261.3824943105.35398>}}]",
  "reason": "{error,{case_clause,etimedout},[{gen_tcp_socket,handle_recv_error,4,[{file,\"gen_tcp_socket.erl\"},{line,2235}]},{gen_statem,loop_state_callback,11,[{file,\"gen_statem.erl\"},{line,1205}]},{proc_lib,init_p_do_apply,3,[{file,\"proc_lib.erl\"},{line,226}]}]}",
  "state": "{{recv,{select_info,recv,#Ref<0.3186442261.3824943105.35398>}},{{params,{'$socket',#Ref<0.3186442261.3805675525.237627>},<0.2154.0>,#Ref<0.3186442261.3805544453.237626>},#{active => true,buffer => <<>>,delay_send => false,deliver => term,exit_on_close => true,header => 0,line_delimiter => 10,mode => binary,packet => raw,packet_size => 67108864,recv_length => 0,send_timeout => 50,send_timeout_close => true,show_econnreset => false,start_opts => [],type => connect}}}",
  "state_enter": false,
  "time": "2023-08-30T10:10:47.793319Z",
  "timeouts": "{0,[]}"
}
```

Socket option was:
```
{socket_options,
                        [binary,
                         {packet,raw},
                         {send_timeout,50},
                         {send_timeout_close,true}]}
```

**Comment by x0id** [comment]
One way to reproduce this error is to create a TCP endpoint which we can connect, but can't receive an ACK for the TCP data sent. I was able to achieve this by adding the following iptables rule serverside:
```
iptables -I OUTPUT -p tcp --sport 8888 --tcp-flags ALL ACK -j DROP
```
It prevented data ACKs from the endpoint running as follows:
```
nc -l 8888
```
Then, straightforward code hit the error:
```
    {ok, Socket} = gen_tcp:connect(Host, 8888, [{debug, true}]),
    gen_tcp:send(Socket, <<"test">>),
    receive
        Msg ->
            io:format("~nmsg: ~p~n~n", [Msg])
    end,
    gen_tcp:close(Socket).
```
This code works with no error with legacy inet backend:
```
INET-DRV-DBG[6794][36,<0.9.0>] inet_set_opts -> opt: 22
INET-DRV-DBG[6838][36,<0.9.0>] inet_set_opts(active) -> active (ival): active (1)
INET-DRV-DBG[11640][36,<0.9.0>] tcp_inet_ctl -> CONNECT
INET-DRV-DBG[11673][36,<0.9.0>] tcp_inet_ctl(connect) -> would block => select
INET-DRV-DBG[11481][36] tcp_inet_stop -> entry
INET-DRV-DBG[10470][36] inet_stop -> entry

msg: {tcp_error,#Port<0.4>,etimedout}

```
But, when run with `-kernel inet_backend socket`, it crashes:
```
SOCKET [02-Sep-2023::00:04:12.109089] [unknown] esock_setopt_otp_debug {36} -> ok
   eVal: true
SOCKET [02-Sep-2023::00:04:12.109162] [unknown] esock_setopt_otp_meta {36} -> entry
   eVal: #{packet=>raw,packet_size=>67108864,delay_send=>false,send_timeout=>infinity,send_timeout_close=>false,show_econnreset=>false}
SOCKET [02-Sep-2023::00:04:12.109294] [unknown] esock_setopt_otp_meta {36} -> ok
SOCKET [02-Sep-2023::00:04:12.109357] [unknown] nif_connect(#Ref<0.1817455145.1021706245.220672>), {360x0} ->
   ConnRef:  #Ref<0.1817455145.1021575173.220675>
   SockAddr: #{port=>8888,addr=>{192,168,1,238},family=>inet}
UNIX-ESSIO [02-Sep-2023::00:04:12.109583] [unknown] essio_connect {36} -> would block => select
SOCKET [02-Sep-2023::00:04:12.109651] [unknown] esock_monitor {36} [<0.84.0>] essio_connect -> conn: try monitor
SOCKET [02-Sep-2023::00:04:12.109718] [unknown] esock_monitor {36} [<0.84.0>] essio_connect -> conn: monitor ok: #Ref<0.1817455145.1021575173.220676>
SOCKET [02-Sep-2023::00:04:12.109982] [unknown] nif_connect(#Ref<0.1817455145.1021706245.220672>) -> done with
   res: select
SOCKET [02-Sep-2023::00:04:12.114272] [unknown] nif_connect(#Ref<0.1817455145.1021706245.220672>), {360x110} ->
   ConnRef:  #Ref<0.1817455145.1021575173.220677>
   SockAddr: #{port=>8888,addr=>{192,168,1,238},family=>inet}
SOCKET [02-Sep-2023::00:04:12.114411] [unknown] esock_demonitor {36} [<0.84.0>] essio_connect finalize -> connected: try demonitor #Ref<0.1817455145.1021575173.220676>
SOCKET [02-Sep-2023::00:04:12.114515] [unknown] nif_connect(#Ref<0.1817455145.1021706245.220672>) -> done with
   res: ok
SOCKET [02-Sep-2023::00:04:12.114590] [unknown] nif_recv(#Ref<0.1817455145.1021706245.220672>), {36,0x0} ->
   recvRef: #Ref<0.1817455145.1021575173.220678>
   len:     0
   flags:   0x0
UNIX-ESSIO [02-Sep-2023::00:04:12.114703] [unknown] essio_recv {36} -> entry with
   count,size: (0:0:8192)
UNIX-ESSIO [02-Sep-2023::00:04:12.114768] [unknown] essio_recv {36} -> try read (8192)
UNIX-ESSIO [02-Sep-2023::00:04:12.114821] [unknown] essio_recv {36} -> read: -1 (35)
UNIX-ESSIO [02-Sep-2023::00:04:12.114872] [unknown] recv_check_result(#Ref<0.1817455145.1021706245.220672>) {36} -> entry with
   read:      -1
   toRead:    0
   saveErrno: 35
   recvRef:   #Ref<0.1817455145.1021575173.220678>
UNIX-ESSIO [02-Sep-2023::00:04:12.115037] [unknown] recv_check_fail(#Ref<0.1817455145.1021706245.220672>) {36} -> eagain
   recvRef: #Ref<0.1817455145.1021575173.220678>
SOCKET [02-Sep-2023::00:04:12.115141] [unknown] esock_monitor {36} [<0.84.0>] recv_init_current_reader -> current reader: try monitor
SOCKET [02-Sep-2023::00:04:12.115225] [unknown] esock_monitor {36} [<0.84.0>] recv_init_current_reader -> current reader: monitor ok: #Ref<0.1817455145.1021575173.220679>
UNIX-ESSIO [02-Sep-2023::00:04:12.115337] [unknown] recv_check_retry(#Ref<0.1817455145.1021706245.220672>) {36} -> SELECT for more
   recvRef: #Ref<0.1817455145.1021575173.220678>
SOCKET [02-Sep-2023::00:04:12.115453] [unknown] nif_recv(#Ref<0.1817455145.1021706245.220672>) -> done
SOCKET [02-Sep-2023::00:04:12.116567] [unknown] esock_getopt_otp -> entry with
   eOpt: 1009
SOCKET [02-Sep-2023::00:04:12.116704] [unknown] esock_getopt_otp_meta {36} ->
   eVal: #{packet=>raw,packet_size=>67108864,delay_send=>false,send_timeout=>infinity,send_timeout_close=>false,show_econnreset=>false}
SOCKET [02-Sep-2023::00:04:12.116844] [unknown] nif_send(#Ref<0.1817455145.1021706245.220672>), {36,0x120} ->
   SendRef:   #Ref<0.1817455145.1021575173.220683>
   Data size: 4
   flags:     0x0
UNIX-ESSIO [02-Sep-2023::00:04:12.116976] [unknown] send_check_result(#Ref<0.1817455145.1021706245.220672>) {36} -> entry with
   send_result:  4
   dataSize:     4
   err:          0
   sendRef:      #Ref<0.1817455145.1021575173.220683>
UNIX-ESSIO [02-Sep-2023::00:04:12.117120] [unknown] send_check_ok(#Ref<0.1817455145.1021706245.220672>) {36} -> everything written (4) - done
SOCKET [02-Sep-2023::00:04:12.117203] [unknown] esock_activate_next_writer(#Ref<0.1817455145.1021706245.220672>) {36} -> no more requestors
SOCKET [02-Sep-2023::00:04:12.117286] [unknown] esock_activate_next_writer(#Ref<0.1817455145.1021706245.220672>) {36} -> done with false
UNIX-ESSIO [02-Sep-2023::00:04:12.117367] [unknown] send_check_ok(#Ref<0.1817455145.1021706245.220672>) {36} -> no more writers
UNIX-ESSIO [02-Sep-2023::00:04:12.117443] [unknown] send_check_result(#Ref<0.1817455145.1021706245.220672>) {36} -> done:
   res: ok
SOCKET [02-Sep-2023::00:04:12.117523] [unknown] nif_send(#Ref<0.1817455145.1021706245.220672>) -> done with
   res: ok
SOCKET [02-Sep-2023::00:05:05.941539] [unknown] nif_recv(#Ref<0.1817455145.1021706245.220672>), {36,0x100} ->
   recvRef: #Ref<0.1817455145.1021575173.220684>
   len:     0
   flags:   0x0
UNIX-ESSIO [02-Sep-2023::00:05:05.942147] [unknown] essio_recv {36} -> entry with
   count,size: (0:0:8192)
UNIX-ESSIO [02-Sep-2023::00:05:05.942467] [unknown] essio_recv {36} -> try read (8192)
UNIX-ESSIO [02-Sep-2023::00:05:05.942688] [unknown] essio_recv {36} -> read: -1 (60)
UNIX-ESSIO [02-Sep-2023::00:05:05.942891] [unknown] recv_check_result(#Ref<0.1817455145.1021706245.220672>) {36} -> entry with
   read:      -1
   toRead:    0
   saveErrno: 60
   recvRef:   #Ref<0.1817455145.1021575173.220684>
UNIX-ESSIO [02-Sep-2023::00:05:05.943457] [unknown] recv_check_fail(#Ref<0.1817455145.1021706245.220672>) {36} -> errno: 60

   recvRef: #Ref<0.1817455145.1021575173.220684>
SOCKET [02-Sep-2023::00:05:05.944026] [unknown] esock_demonitor {36} [<0.84.0>] recv_error_current_reader: try demonitor #Ref<0.1817455145.1021575173.220679>
SOCKET [02-Sep-2023::00:05:05.944553] [unknown] nif_recv(#Ref<0.1817455145.1021706245.220672>) -> done
SOCKET [02-Sep-2023::00:05:05.945062] [unknown] nif_close(#Ref<0.1817455145.1021706245.220672>) ->
      Socket:      36
      Read State:  0x100
      Write State: 0x120
      Caller:      <0.84.0>
UNIX-ESSIO [02-Sep-2023::00:05:05.945635] [unknown] essio_close {36} -> stop was called
SOCKET [02-Sep-2023::00:05:05.945853] [unknown] nif_close(#Ref<0.1817455145.1021706245.220672>) -> done
   res: ok
SOCKET [02-Sep-2023::00:05:05.946680] [unknown] nif_finalize_close(#Ref<0.1817455145.1021706245.220672>, 36) -> 
   ReadState:  0x300
   WriteState: 0x320
SOCKET [02-Sep-2023::00:05:05.947134] [unknown] esock_demonitor {36} [<0.84.0>] essio_fin_close -> ctrl: try demonitor #Ref<0.1817455145.1021575173.220673>
SOCKET [02-Sep-2023::00:05:05.947513] [unknown] esock_close_socket(36) -> try socket close
SOCKET [02-Sep-2023::00:05:05.947777] [unknown] nif_finalize_close(#Ref<0.1817455145.1021706245.220672>) -> done with
   result: ok
UNIX-ESSIO [02-Sep-2023::00:05:05.948150] [unknown] essio_cancel_recv(#Ref<0.1817455145.1021706245.220672>), {-1,0x700} -> entry with
   opRef: #Ref<0.1817455145.1021575173.220678>
   without reader
UNIX-ESSIO [02-Sep-2023::00:05:05.948568] [unknown] essio_cancel_recv(#Ref<0.1817455145.1021706245.220672>) {-1} -> done with result:
   {error,closed}

msg: {tcp_error,{'$inet',gen_tcp_socket,
                         {<0.84.0>,
                          {'$socket',#Ref<0.1817455145.1021706245.220672>}}},
                etimedout}

=ERROR REPORT==== 2-Sep-2023::00:05:05.958466 ===
** State machine <0.84.0> terminating
** Last event = {info,
                    {'$socket',
                        {'$socket',#Ref<0.1817455145.1021706245.220672>},
                        select,#Ref<0.1817455145.1021575173.220678>}}
** When server state  = {{recv,
                             {select_info,recv,
                                 #Ref<0.1817455145.1021575173.220678>}},
                         {{params,
                              {'$socket',#Ref<0.1817455145.1021706245.220672>},
                              <0.9.0>,#Ref<0.1817455145.1021575176.220908>},
                          #{active => true,line_delimiter => 10,packet => raw,
                            packet_size => 67108864,type => connect,
                            mode => list,buffer => <<>>,delay_send => false,
                            deliver => term,exit_on_close => true,header => 0,
                            send_timeout => infinity,
                            send_timeout_close => false,
                            show_econnreset => false,start_opts => [],
                            recv_length => 0}}}
** Reason for termination = error:{case_clause,etimedout}
** Callback modules = [gen_tcp_socket]
** Callback mode = handle_event_function
** Stacktrace =
**  [{gen_tcp_socket,handle_recv_error,4,
                     [{file,"gen_tcp_socket.erl"},{line,2250}]},
     {gen_statem,loop_state_callback,11,[{file,"gen_statem.erl"},{line,1379}]},
     {proc_lib,init_p_do_apply,3,[{file,"proc_lib.erl"},{line,241}]}]

=CRASH REPORT==== 2-Sep-2023::00:05:05.959112 ===
  crasher:
    initial call: gen_tcp_socket:init/1
    pid: <0.84.0>
    registered_name: []
    exception error: no case clause matching etimedout
      in function  gen_tcp_socket:handle_recv_error/4 (gen_tcp_socket.erl, line 2250)
      in call from gen_statem:loop_state_callback/11 (gen_statem.erl, line 1379)
    ancestors: [<0.9.0>]
    message_queue_len: 1
    messages: [{'$gen_call',{<0.9.0>,#Ref<0.1817455145.1021575176.220914>},
                               close}]
    links: []
    dictionary: []
    trap_exit: true
    status: running
    heap_size: 6772
    stack_size: 28
    reductions: 11980
  neighbours:

=ERROR REPORT==== 2-Sep-2023::00:05:05.962035 ===
gen_tcp_socket call failed: 
      Call:  close
      Class: exit
      Error: {{{case_clause,etimedout},
               [{gen_tcp_socket,handle_recv_error,4,
                                [{file,"gen_tcp_socket.erl"},{line,2250}]},
                {gen_statem,loop_state_callback,11,
                            [{file,"gen_statem.erl"},{line,1379}]},
                {proc_lib,init_p_do_apply,3,
                          [{file,"proc_lib.erl"},{line,241}]}]},
              {gen_statem,call,[<0.84.0>,close,infinity]}}
      Stack: [{gen,do_call,4,[{file,"gen.erl"},{line,246}]},
              {gen_statem,call,3,[{file,"gen_statem.erl"},{line,905}]},
              {gen_tcp_socket,call,2,
                              [{file,"gen_tcp_socket.erl"},{line,1143}]},
              {gen_tcp_socket,close_server,1,
                              [{file,"gen_tcp_socket.erl"},{line,475}]},
              {gen_tcp_socket,close,1,
                              [{file,"gen_tcp_socket.erl"},{line,470}]},
              {erl_eval,local_func,8,[{file,"erl_eval.erl"},{line,646}]},
              {escript,interpret,4,[{file,"escript.erl"},{line,780}]},
              {escript,start,1,[{file,"escript.erl"},{line,277}]}]

escript: exception exit: {{{case_clause,etimedout},
                  [{gen_tcp_socket,handle_recv_error,4,
                                   [{file,"gen_tcp_socket.erl"},{line,2250}]},
                   {gen_statem,loop_state_callback,11,
                               [{file,"gen_statem.erl"},{line,1379}]},
                   {proc_lib,init_p_do_apply,3,
                             [{file,"proc_lib.erl"},{line,241}]}]},
                 {gen_statem,call,[<0.84.0>,close,infinity]}}
  in function  gen:do_call/4 (gen.erl, line 246)
  in call from gen_statem:call/3 (gen_statem.erl, line 905)
  in call from gen_tcp_socket:call/2 (gen_tcp_socket.erl, line 1143)
  in call from gen_tcp_socket:close_server/1 (gen_tcp_socket.erl, line 475)
  in call from gen_tcp_socket:close/1 (gen_tcp_socket.erl, line 470)
```

**Comment by RaimoNiskanen** [comment]
Since it seems like a hazel to reproduce this, maybe you can try PR #7626 to see if it fixes the problem. :-)

**Comment by x0id** [comment]
> Since it seems like a hazel to reproduce this
Why it's hassle? I described an idea how to do it :)

**Comment by x0id** [comment]
@IngelaAndin In order to test the patch I need to install the branched version of OTP on my side. I do not know how to do this since I'm not an active contributor to this project. If you could help me navigate the steps, I can try.
But for the sake of truth, I don't agree that testing the steps I described involves any administrative changes. It might be done with docker-compose. Of course, it requires some time and technical expertise to implement a test.

**Comment by IngelaAndin** [comment]
@x0id  just clone the repo and build some help can be found at: https://github.com/erlang/otp

**Comment by x0id** [comment]
> @x0id just clone the repo and build some help can be found at: https://github.com/erlang/otp

FYI: the better way to build it is to use kerl:
```
kerl build git https://github.com/RaimoNiskanen/otp.git \
"raimo/kernel/gen_tcp_socket-etimedout/GH-7608/OTP-18749" fix

kerl install fix ~/kerl/fix

. ~/kerl/fix/activate
```

Test went ok.

**Comment by IngelaAndin** [comment]
@x0id  Thank you for testing!  How to best build might be a little subjective but what ever works best for you :)

**Comment by x0id** [comment]
> @x0id Thank you for testing! How to best build might be a little subjective but what ever works best for you :)

For me, it would be better not to build it at all, but I didn't expect the owner of the project to not be capable of testing their product. Not a frequent case for the times when I grew as a professional...

**Comment by IngelaAndin** [comment]
Of course we are able to test our product under normal circumstances.  This is  kind of special circumstances.  We also often ask users to verify things in their system even when we have very good local test and we are 99.99 % sure it will also fix their problem. We will  also never not make a test case for something if it is feasible to do so in an automated way with reasonable effort. Now that we know the patch solves your problem we can try to figure out if there maybe is some other way to tests it.

**Comment by RaimoNiskanen** [comment]
Merged to 'maint' to be released in OTP-26.2.

Thank you for reporting this issue!
